### ISSUE TITLE
Request and debug not installed with NPM, where are environment variables located?

### ISSUE THREAD

**Issue by webdevbrian** [alaycock/sirius-playlists#4]
Hi, noticed running an npm install doesn't pull these two required libs in from the base directory.

Also, where are the environment variables located so I can supply my own YT_API_KEY?

Would love to modify this in a fork and fix / redo the UI, just need to get this running, and I'll add some PRs here to give back.

Thanks!

**Comment by alaycock** [comment]
For sure. Let me update the docs a bit to clear this up. This was a personal project and I didn't expect anyone else to contribute or install, but I'm glad you are!

**Comment by webdevbrian** [comment]
Hell yeah man! I'll totally help out. let me know!

**Comment by webdevbrian** [comment]
Yeah we'll probably be going back and forth a few times ;)

>  /sirius-playlists/node_modules/monk/lib/manager.js:29
    throw Error('No connection URI provided.')

I did create a `.env` file in the project's base directory

**Comment by webdevbrian** [comment]
Actually running it with `foreman start` seemingly made me get further without that error, but now I have a new one which is:

>  Error: No default engine was specified and no extension was provided.

**Comment by alaycock** [comment]
Yeah, the monk issue is caused by the environment variable not being read from `.env`. If you start it with `npm start`, it won't read `.env`, but if you use `foreman start`, it will, so you did the right thing.

Not sure why you're getting the `no default engine` error. Try adding `app.set('view engine', 'jade');` onto line 22 of app.js and let me know if that fixes it.

**Comment by webdevbrian** [comment]
Getting closer now, had to install jade, and add an if errorcheck in the error.jade file. I might honestly want to forego jade for this, I could actually through bootstrap in there, so if you want to branch it and remove that so I can do that, that'd actually be hugely beneficial.

Jade is actually now pug anyway, if I recall correctly.

Current issue is a 500 from the getsong request on the node 3000 port. Any ideas?

**Comment by alaycock** [comment]
Yeah, I should definitely pull out Jade. I'm pretty sure I don't even use it for anything. Does anything print to the console when you get the 500 error?

I'll try pulling out jade and pushing a new version in a bit.

**Comment by webdevbrian** [comment]
This is what I have, just kind of rolls on along with these:

> 20:24:43 web.1  | GET /getsong?c=8205 500 6.628 ms - 336
> 20:25:13 web.1  | GET /getsong?c=8205 500 7.904 ms - 336
> 20:25:43 web.1  | GET /getsong?c=8205 500 6.477 ms - 336
> 20:26:13 web.1  | GET /getsong?c=8205 500 7.004 ms - 336
> 20:26:43 web.1  | GET /getsong?c=8205 500 7.207 ms - 336


Chrome console:

> Searching queue
> ajax.js:11 8205
> jquery-2.1.1.min.js:4 GET http://localhost:3000/getsong?c=8205 500 (Internal Server Error)

**Comment by alaycock** [comment]
Just pushed another change, it should now log an error message in the console when there's a 500 error. It doesn't do the whole stack trace, but you could change that by modifying line 73 of app.js, if you want more detail.

I also completely pulled out jade, as far as I can tell, so that shouldn't be an issue again, hopefully.

**Comment by webdevbrian** [comment]
Nice, thanks for the jade thing.

>  21:40:13 web.1  | GET / 304 0.639 ms - -
21:40:13 web.1  | GET /stylesheets/style.css 304 0.800 ms - -
21:40:13 web.1  | GET /js/jquery-2.1.1.min.js 304 0.549 ms - -
21:40:13 web.1  | GET /js/ajax.js 304 3.823 ms - -
21:40:13 web.1  | GET /js/channels.js 304 3.415 ms - -
21:40:13 web.1  | Cannot set property 'res' of undefined
21:40:13 web.1  | GET /getsong?c=8205 500 1.110 ms - 38

**Comment by alaycock** [comment]
Pushed up another fix. Sorry for all the little things, I wrote this ages ago and didn't do a great job of polishing it.

We also might be running different node versions, which could explain why you're getting errors when I'm not.

Either way, try out that fix and let me know if it works.

**Comment by webdevbrian** [comment]
Looks like I'm getting a 200 status code in my terminal now with a feed of artists:

>  22:18:36 web.1  |      statusCode: 200,
22:18:36 web.1  |      body: '{"channelMetadataResponse":{"messages":{"code":100,"message":"Successful request"},"status":1,"metaData":{"channelId":"area33","channelName":"Electric Area","channelNumber":52,"currentEvent":{"artists":{"id":"","name":"@Borgore @sxmElectro"},"baseUrl":"http:\\/\\/www.siriusxm.com\\/albumart\\/","epgInfo":{"program":{"description":"The Borgore Show brings you the music that disrupts all sides of the music spectrum. This show represents a whole new level of sound to electronic music by bringing the hottest and freshest beats of Dubstep, House, and Trap to anywhere in the world here on SiriusXM\'s Electric Area!","episodes":{"description":"The Borgore Show brings you the music that disrupts all sides of the music spectrum. This show represents a whole new level of sound to electronic music by bringing the hottest and freshest beats of Dubstep, House, and Trap to anywhere in the world here on SiriusXM\'s Electric Area!","duration":3600,"episodeId":3434103,"firstRun":false,"showId":6187,"startTime":"2017-01-11T22:00:00Z","title":"The Borgore Show"},"firstRun":false,"live":false,"showId":6187,"startTime":"2017-01-11T22:00:00Z","title":"The Borgore Show"},"segment":{"description":"The Borgore Show brings you the music that disrupts all sides of the music spectrum. This show represents a whole new level of sound to electronic music by bringing the hottest and freshest beats of Dubstep, House, and Trap to anywhere in the world here on SiriusXM\'s Electric Area!","duration":3600,"segmentId":3434103,"startTime":"2017-01-11T22:00:00Z"}},"keyIndex":5,"keyValue":"12345678901234567890123456789012","siriusXMId":380896108,"song":{"album":{"name":"(MIX)"},"composer":"","creativeArts":[{"encrypted":false,"size":"THUMBNAIL","type":"IMAGE","url":"Live\\/Default\\/DefaultMDS_t_52.jpg"},{"encrypted":false,"size":"SMALL","type":"IMAGE","url":"Live\\/Default\\/DefaultMDS_s_52.jpg"},{"encrypted":false,"size":"MEDIUM","type":"IMAGE","url":"Live\\/Default\\/DefaultMDS_m_52.jpg"},{"encrypted":false,"size":"LARGE","type":"IMAGE","url":""},{"encrypted":false,"type":"BIO","url":""},{"encrypted":false,"type":"REVIEWS","url":""},{"encrypted":true,"size":"THUMBNAIL","type":"IMAGE","url":"NY_Assets_1\\/Default_Content\\/5_DefaultMDS_t_52.jpg"},{"encrypted":true,"size":"SMALL","type":"IMAGE","url":"NY_Assets_1\\/Default_Content\\/5_DefaultMDS_s_52.jpg"},{"encrypted":true,"size":"MEDIUM","type":"IMAGE","url":"NY_Assets_1\\/Default_Content\\/5_DefaultMDS_m_52.jpg"},{"encrypted":true,"size":"LARGE","type":"IMAGE","url":""},{"encrypted":true,"type":"BIO","url":""},{"encrypted":true,"type":"REVIEWS","url":""}],"id":"","name":"#TheBorgoreShow"},"startTime":"2017-01-12T03:02:02Z"},"dateTime":"2017-01-12T03:18:02.745Z","version":1.1}}}' } }

Chrome console is now spitting this out however:

>  Object {error: "Could not get song from YouTube."}
ajax.js:35 No song to push to queue. Server error, trying again in 30 seconds
ajax.js:40 []
ajax.js:52 Could not setup player, trying again in 30 seconds
ajax.js:8 Searching queue
ajax.js:11 area33

**Comment by alaycock** [comment]
Okay, this looks like everything is working as expected, but this is the one issue with how this program works. Basically it finds the song on Sirius XM, then finds the best matching song on youtube. If you look at the JSON response for these two keys:

```
data.channelMetadataResponse.metaData.currentEvent.artists.name
data.channelMetadataResponse.metaData.currentEvent.song.name
```

That's where the artist name and song name are stored, and in your JSON result, the artist is `@Borgore @sxmElectro` and the song name is `(MIX)`. That means that right now there is someone playing a mix and my scraper can't pull the song name, since it has no idea what song it is since it's a mix.

It's kind of an unfortunate situation, since there's no way to know what is playing. So sometimes some stations just won't be able to play. Other stations should still work though.

**Comment by webdevbrian** [comment]
Hmm, every other station is doing the same actually. Quite odd. No player is present  on the local 3000 port, just the channel list, shown here:

![screenshot on 2017-01-11 at 22-49-44](https://cloud.githubusercontent.com/assets/308749/21876370/49805c50-d850-11e6-9b9d-190ec9cf23db.png)

and for the record, the `.env` file should be located here, correct?:

![screenshot on 2017-01-11 at 22-50-31](https://cloud.githubusercontent.com/assets/308749/21876392/6d602c40-d850-11e6-94be-20b2e420069b.png)

**Comment by alaycock** [comment]
Yep, the `.env` is in the right place. 

It might be a youtube error, I added another logging statement to print the youtube error so that might help track down the issue.

**Comment by webdevbrian** [comment]
Huzzah, getting further :) Came down to access with granting the app I had to setup with youtube and having API access. Doh!

>  23:51:19 web.1  |     if(item.id.kind == "youtube#video") {
23:51:19 web.1  |            ^
23:51:19 web.1  |
23:51:19 web.1  | TypeError: Cannot read property 'id' of undefined

First song plays now, but then it throws up and dies after the above

**Comment by webdevbrian** [comment]
Strike that, got it working. Sweet, thanks for the help.  Going to start on a few wireframes and see where I can go with this now.

**Comment by webdevbrian** [comment]
I actually retract that, I'm getting that same error on one of my other laptops:

> 
> 23:51:19 web.1 | if(item.id.kind == "youtube#video") {
> 23:51:19 web.1 | ^
> 23:51:19 web.1 |
> 23:51:19 web.1 | TypeError: Cannot read property 'id' of undefined

**Comment by alaycock** [comment]
I pushed another change that should help prevent that error, let me know if that helps.

**Comment by webdevbrian** [comment]
Thanks!
